hack/licence-detector/dependency/dependency.go

```diff
@@ -0,0 +1,80 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package dependency
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+)
+
+// List holds direct and indirect dependency information.
+type List struct {
+	Direct   []Info
+	Indirect []Info
+}
+
+// Info holds information about a dependency.
+type Info struct {
+	Name            string `json:"name"`
+	Dir             string `json:"-"`
+	LicenceFile     string `json:"-"`
+	LicenceType     string `json:"licenceType"`
+	URL             string `json:"url"`
+	Version         string `json:"version"`
+	VersionTime     string `json:"versionTime"`
+	LicenceTextFile string `json:"licenceTextFile"`
+}
+
+// Overrides is a mapping from module name to dependency info.
+type Overrides map[string]Info
+
+// LoadOverrides loads the dependency overrides from the given file.
+// LicenceTextFile will be read relative to the parent directory of the file.
+func LoadOverrides(file string) (Overrides, error) {
+	depMap := make(Overrides)
+	if file == "" {
+		return depMap, nil
+	}
+
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open overrides file %s: %w", file, err)
+	}
+	defer f.Close()
+
+	rootDir, err := filepath.Abs(filepath.Dir(file))
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine absolute path of overrides file: %w", err)
+	}
+
+	decoder := json.NewDecoder(f)
+	for {
```

Comment:
This might be me overthinking it, but we mostly parse it this way because we want to end up with a map for `mkDepInfo()`, correct? If we wanted a slice we could just store overrides.json as a json array and unmarshal it right into a slice (and also get to prettify the JSON file rather than making each item be on one line).

Reply:
I chose the JSON lines format because it is well-known and Go supports it out of the box.
The amount of data stored on each line is pretty minimal and I don't expect it to be edited too often. If it starts to become too complex and requires pretty formatting, I think we can consider switching to a more human-friendly file format at that point.